File: nop/ts/admin/src/protocol/wire.ts
// This file is part of the product NoPressure.
// SPDX-FileCopyrightText: 2025-2026 Zivatar Limited
// SPDX-License-Identifier: AGPL-3.0-or-later
// The code and documentation in this repository is licensed under the GNU Affero General Public License v3.0 or later (AGPL-3.0-or-later). See LICENSE.

export class WireError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "WireError";
  }
}

export interface WireEncode {
  encode(writer: WireWriter): void;
}

export interface WireDecode<T> {
  decode(reader: WireReader): T;
}

const textEncoder = new TextEncoder();
const textDecoder = new TextDecoder("utf-8");

export class WireWriter {
  private chunks: Uint8Array[] = [];
  private length = 0;

  writeU8(value: number): void {
    const buf = new Uint8Array(1);
    buf[0] = value & 0xff;
    this.push(buf);
  }

  writeU16(value: number): void {
    const buf = new Uint8Array(2);
    const view = new DataView(buf.buffer);
    view.setUint16(0, value & 0xffff, true);
    this.push(buf);
  }

  writeU32(value: number): void {
    const buf = new Uint8Array(4);
    const view = new DataView(buf.buffer);
    view.setUint32(0, value >>> 0, true);
    this.push(buf);
  }

  writeU64(value: number | bigint): void {
    const big = this.toU64(value);
    const buf = new Uint8Array(8);
    const view = new DataView(buf.buffer);
    const low = Number(big & 0xffff_ffffn);
    const high = Number((big >> 32n) & 0xffff_ffffn);
    view.setUint32(0, low >>> 0, true);
    view.setUint32(4, high >>> 0, true);
    this.push(buf);
  }

  writeI32(value: number): void {
    const buf = new Uint8Array(4);
    const view = new DataView(buf.buffer);
    view.setInt32(0, value | 0, true);
    this.push(buf);
  }

  writeBool(value: boolean): void {
    this.writeU8(value ? 1 : 0);
  }

  writeBytes(bytes: Uint8Array): void {
    this.writeU32(bytes.length);
    this.push(bytes);
  }

  writeString(value: string): void {
    const bytes = textEncoder.encode(value);
    this.writeBytes(bytes);
  }

  writeVec<T>(values: readonly T[], writeItem: (writer: WireWriter, value: T) => void): void {
    if (values.length > 0xffff_ffff) {
      throw new WireError("Vector length exceeds u32 limit");
    }
    this.writeU32(values.length);
    for (const value of values) {
      writeItem(this, value);
    }
  }

  toUint8Array(): Uint8Array {
    const result = new Uint8Array(this.length);
    let offset = 0;
    for (const chunk of this.chunks) {
      result.set(chunk, offset);
      offset += chunk.length;
    }
    return result;
  }

  private push(bytes: Uint8Array): void {
    if (bytes.length === 0) {
      return;
    }
    this.chunks.push(bytes);
    this.length += bytes.length;
  }

  private toU64(value: number | bigint): bigint {
    if (typeof value === "number") {
      if (!Number.isSafeInteger(value) || value < 0) {
        throw new WireError("Invalid u64 number value");
      }
      return BigInt(value);
    }
    if (value < 0n || value > 0xffff_ffff_ffff_ffffn) {
      throw new WireError("Invalid u64 bigint value");
    }
    return value;
  }
}

export class WireReader {
  private view: DataView;
  private offset = 0;

  constructor(private bytes: Uint8Array) {
    this.view = new DataView(
      bytes.buffer,
      bytes.byteOffset,
      bytes.byteLength,
    );
  }

  readU8(): number {
    this.ensureAvailable(1);
    const value = this.view.getUint8(this.offset);
    this.offset += 1;
    return value;
  }

  readU16(): number {
    this.ensureAvailable(2);
    const value = this.view.getUint16(this.offset, true);
    this.offset += 2;
    return value;
  }

  readU32(): number {
    this.ensureAvailable(4);
    const value = this.view.getUint32(this.offset, true);
    this.offset += 4;
    return value;
  }

  readU64(): bigint {
    this.ensureAvailable(8);
    const low = BigInt(this.view.getUint32(this.offset, true));
    const high = BigInt(this.view.getUint32(this.offset + 4, true));
    this.offset += 8;
    return (high << 32n) | low;
  }

  readU64Number(): number {
    const value = this.readU64();
    if (value > BigInt(Number.MAX_SAFE_INTEGER)) {
      throw new WireError("Decoded u64 exceeds safe integer range");
    }
    return Number(value);
  }

  readI32(): number {
    this.ensureAvailable(4);
    const value = this.view.getInt32(this.offset, true);
    this.offset += 4;
    return value;
  }

  readBool(): boolean {
    const value = this.readU8();
    if (value === 0) {
      return false;
    }
    if (value === 1) {
      return true;
    }
    throw new WireError("Invalid boolean value");
  }

  readBytes(): Uint8Array {
    const length = this.readU32();
    this.ensureAvailable(length);
    const slice = this.bytes.subarray(this.offset, this.offset + length);
    this.offset += length;
    return slice;
  }

  readString(): string {
    const bytes = this.readBytes();
    return textDecoder.decode(bytes);
  }

  readVec<T>(readItem: (reader: WireReader) => T): T[] {
    const length = this.readU32();
    const values: T[] = [];
    for (let i = 0; i < length; i += 1) {
      values.push(readItem(this));
    }
    return values;
  }

  ensureFullyConsumed(): void {
    if (this.offset !== this.bytes.length) {
      throw new WireError("Trailing bytes after decode");
    }
  }

  private ensureAvailable(length: number): void {
    if (this.offset + length > this.bytes.length) {
      throw new WireError("Unexpected end of buffer");
    }
  }
}

type OptionWidth = "u8" | "u16" | "u32" | "u64";

function optionWidthForCount(count: number): OptionWidth {
  if (count <= 0) {
    throw new WireError("Option map requires at least one field");
  }
  if (count <= 8) {
    return "u8";
  }
  if (count <= 16) {
    return "u16";
  }
  if (count <= 32) {
    return "u32";
  }
  if (count <= 64) {
    return "u64";
  }
  throw new WireError("Option map exceeds 64 fields");
}

export class OptionMap {
  static write(writer: WireWriter, flags: readonly boolean[]): void {
    const width = optionWidthForCount(flags.length);
    let value = 0n;
    flags.forEach((enabled, index) => {
      if (enabled) {
        value |= 1n << BigInt(index);
      }
    });
    switch (width) {
      case "u8":
        writer.writeU8(Number(value));
        break;
      case "u16":
        writer.writeU16(Number(value));
        break;
      case "u32":
        writer.writeU32(Number(value));
        break;
      case "u64":
        writer.writeU64(value);
        break;
      default:
        throw new WireError("Unsupported option map width");
    }
  }

  static read(reader: WireReader, count: number): boolean[] {
    if (count === 0) {
      return [];
    }
    const width = optionWidthForCount(count);
    let value: bigint;
    switch (width) {
      case "u8":
        value = BigInt(reader.readU8());
        break;
      case "u16":
        value = BigInt(reader.readU16());
        break;
      case "u32":
        value = BigInt(reader.readU32());
        break;
      case "u64":
        value = reader.readU64();
        break;
      default:
        throw new WireError("Unsupported option map width");
    }

    const mask =
      count === 64 ? (1n << 64n) - 1n : (1n << BigInt(count)) - 1n;
    if ((value & ~mask) !== 0n) {
      throw new WireError("Option map contains unknown bits");
    }

    const flags: boolean[] = [];
    for (let index = 0; index < count; index += 1) {
      flags.push(((value >> BigInt(index)) & 1n) === 1n);
    }
    return flags;
  }
}

export class WorkflowCounter {
  private nextId = 1;

  next(): number {
    if (this.nextId === 0) {
      throw new WireError("Workflow counter wrapped");
    }
    if (this.nextId === 0xffff_ffff) {
      throw new WireError("Workflow counter exhausted");
    }
    const id = this.nextId;
    this.nextId += 1;
    return id;
  }
}

export class WorkflowTracker {
  private lastId = 0;

  accept(workflowId: number): void {
    if (workflowId === 0) {
      throw new WireError("Workflow ID must be non-zero");
    }
    if (workflowId <= this.lastId) {
      throw new WireError("Workflow ID must be strictly increasing");
    }
    this.lastId = workflowId;
  }
}

File: nop/src/management/wire.rs
// This file is part of the product NoPressure.
// SPDX-FileCopyrightText: 2025-2026 Zivatar Limited
// SPDX-License-Identifier: AGPL-3.0-or-later
// The code and documentation in this repository is licensed under the GNU Affero General Public License v3.0 or later (AGPL-3.0-or-later). See LICENSE.

use std::error::Error;
use std::fmt;

pub type WireResult<T> = Result<T, WireError>;

#[derive(Debug)]
pub struct WireError {
    message: String,
}

impl WireError {
    pub fn new(message: impl Into<String>) -> Self {
        Self {
            message: message.into(),
        }
    }
}

impl fmt::Display for WireError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "wire error: {}", self.message)
    }
}

impl Error for WireError {}

pub trait WireEncode {
    fn encode(&self, writer: &mut WireWriter) -> WireResult<()>;
}

pub trait WireDecode: Sized {
    fn decode(reader: &mut WireReader) -> WireResult<Self>;
}

#[derive(Debug, Clone, Copy)]
pub enum OptionWidth {
    U8,
    U16,
    U32,
    U64,
}

impl OptionWidth {
    fn for_count(count: usize) -> WireResult<Self> {
        match count {
            0 => Err(WireError::new("Option map requires at least one field")),
            1..=8 => Ok(OptionWidth::U8),
            9..=16 => Ok(OptionWidth::U16),
            17..=32 => Ok(OptionWidth::U32),
            33..=64 => Ok(OptionWidth::U64),
            _ => Err(WireError::new("Option map exceeds 64 fields")),
        }
    }
}

#[derive(Debug, Clone, Copy)]
pub struct OptionMap {
    width: OptionWidth,
    value: u64,
}

impl OptionMap {
    pub fn from_flags(flags: &[bool]) -> WireResult<Self> {
        let width = OptionWidth::for_count(flags.len())?;
        let mut value = 0u64;
        for (index, enabled) in flags.iter().enumerate() {
            if *enabled {
                value |= 1u64 << index;
            }
        }
        Ok(Self { width, value })
    }

    pub fn write(&self, writer: &mut WireWriter) -> WireResult<()> {
        match self.width {
            OptionWidth::U8 => writer.write_u8(self.value as u8),
            OptionWidth::U16 => writer.write_u16(self.value as u16),
            OptionWidth::U32 => writer.write_u32(self.value as u32),
            OptionWidth::U64 => writer.write_u64(self.value),
        }
        Ok(())
    }

    pub fn read(reader: &mut WireReader, count: usize) -> WireResult<Vec<bool>> {
        if count == 0 {
            return Ok(Vec::new());
        }
        let width = OptionWidth::for_count(count)?;
        let value = match width {
            OptionWidth::U8 => reader.read_u8()? as u64,
            OptionWidth::U16 => reader.read_u16()? as u64,
            OptionWidth::U32 => reader.read_u32()? as u64,
            OptionWidth::U64 => reader.read_u64()?,
        };

        let mask = if count == 64 {
            u64::MAX
        } else {
            (1u64 << count) - 1
        };
        if value & !mask != 0 {
            return Err(WireError::new("Option map contains unknown bits"));
        }

        let mut flags = Vec::with_capacity(count);
        for index in 0..count {
            flags.push((value >> index) & 1 == 1);
        }
        Ok(flags)
    }
}

#[derive(Debug, Default)]
pub struct WireWriter {
    buffer: Vec<u8>,
}

impl WireWriter {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn into_bytes(self) -> Vec<u8> {
        self.buffer
    }

    pub fn write_u8(&mut self, value: u8) {
        self.buffer.push(value);
    }

    pub fn write_u16(&mut self, value: u16) {
        self.buffer.extend_from_slice(&value.to_le_bytes());
    }

    pub fn write_u32(&mut self, value: u32) {
        self.buffer.extend_from_slice(&value.to_le_bytes());
    }

    pub fn write_u64(&mut self, value: u64) {
        self.buffer.extend_from_slice(&value.to_le_bytes());
    }

    pub fn write_i32(&mut self, value: i32) {
        self.buffer.extend_from_slice(&value.to_le_bytes());
    }

    pub fn write_bool(&mut self, value: bool) {
        self.write_u8(if value { 1 } else { 0 });
    }

    pub fn write_bytes(&mut self, bytes: &[u8]) -> WireResult<()> {
        let len = u32::try_from(bytes.len())
            .map_err(|_| WireError::new("Byte slice exceeds u32 length limit"))?;
        self.write_u32(len);
        self.buffer.extend_from_slice(bytes);
        Ok(())
    }

    pub fn write_string(&mut self, value: &str) -> WireResult<()> {
        self.write_bytes(value.as_bytes())
    }

    pub fn write_vec<T, F>(&mut self, values: &[T], mut write_item: F) -> WireResult<()>
    where
        F: FnMut(&mut WireWriter, &T) -> WireResult<()>,
    {
        let len = u32::try_from(values.len())
            .map_err(|_| WireError::new("Vector length exceeds u32 limit"))?;
        self.write_u32(len);
        for value in values {
            write_item(self, value)?;
        }
        Ok(())
    }
}

pub struct WireReader<'a> {
    bytes: &'a [u8],
    offset: usize,
}

impl<'a> WireReader<'a> {
    pub fn new(bytes: &'a [u8]) -> Self {
        Self { bytes, offset: 0 }
    }

    pub fn ensure_fully_consumed(&self) -> WireResult<()> {
        if self.offset != self.bytes.len() {
            return Err(WireError::new("Trailing bytes after decode"));
        }
        Ok(())
    }

    pub fn read_u8(&mut self) -> WireResult<u8> {
        self.ensure_available(1)?;
        let value = self.bytes[self.offset];
        self.offset += 1;
        Ok(value)
    }

    pub fn read_u16(&mut self) -> WireResult<u16> {
        self.ensure_available(2)?;
        let value = u16::from_le_bytes([self.bytes[self.offset], self.bytes[self.offset + 1]]);
        self.offset += 2;
        Ok(value)
    }

    pub fn read_u32(&mut self) -> WireResult<u32> {
        self.ensure_available(4)?;
        let value = u32::from_le_bytes([
            self.bytes[self.offset],
            self.bytes[self.offset + 1],
            self.bytes[self.offset + 2],
            self.bytes[self.offset + 3],
        ]);
        self.offset += 4;
        Ok(value)
    }

    pub fn read_u64(&mut self) -> WireResult<u64> {
        self.ensure_available(8)?;
        let value = u64::from_le_bytes([
            self.bytes[self.offset],
            self.bytes[self.offset + 1],
            self.bytes[self.offset + 2],
            self.bytes[self.offset + 3],
            self.bytes[self.offset + 4],
            self.bytes[self.offset + 5],
            self.bytes[self.offset + 6],
            self.bytes[self.offset + 7],
        ]);
        self.offset += 8;
        Ok(value)
    }

    pub fn read_i32(&mut self) -> WireResult<i32> {
        self.ensure_available(4)?;
        let value = i32::from_le_bytes([
            self.bytes[self.offset],
            self.bytes[self.offset + 1],
            self.bytes[self.offset + 2],
            self.bytes[self.offset + 3],
        ]);
        self.offset += 4;
        Ok(value)
    }

    pub fn read_bool(&mut self) -> WireResult<bool> {
        let value = self.read_u8()?;
        match value {
            0 => Ok(false),
            1 => Ok(true),
            _ => Err(WireError::new("Invalid boolean value")),
        }
    }

    pub fn read_bytes(&mut self) -> WireResult<Vec<u8>> {
        let len = self.read_u32()? as usize;
        self.ensure_available(len)?;
        let slice = self.bytes[self.offset..self.offset + len].to_vec();
        self.offset += len;
        Ok(slice)
    }

    pub fn read_string(&mut self) -> WireResult<String> {
        let bytes = self.read_bytes()?;
        String::from_utf8(bytes).map_err(|_| WireError::new("Invalid UTF-8 string"))
    }

    pub fn read_vec<T, F>(&mut self, mut read_item: F) -> WireResult<Vec<T>>
    where
        F: FnMut(&mut WireReader<'a>) -> WireResult<T>,
    {
        let len = self.read_u32()? as usize;
        let mut values = Vec::with_capacity(len);
        for _ in 0..len {
            values.push(read_item(self)?);
        }
        Ok(values)
    }

    fn ensure_available(&self, len: usize) -> WireResult<()> {
        if self.offset + len > self.bytes.len() {
            return Err(WireError::new("Unexpected end of buffer"));
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn wire_writer_reader_roundtrip() {
        let mut writer = WireWriter::new();
        writer.write_u8(7);
        writer.write_u16(500);
        writer.write_u32(42);
        writer.write_u64(123456);
        writer.write_i32(-12);
        writer.write_bool(true);
        writer.write_string("hi").expect("string");
        writer
            .write_vec(&["a", "b"], |w, value| w.write_string(value))
            .expect("vec");

        let bytes = writer.into_bytes();
        let mut reader = WireReader::new(&bytes);
        assert_eq!(reader.read_u8().unwrap(), 7);
        assert_eq!(reader.read_u16().unwrap(), 500);
        assert_eq!(reader.read_u32().unwrap(), 42);
        assert_eq!(reader.read_u64().unwrap(), 123456);
        assert_eq!(reader.read_i32().unwrap(), -12);
        assert!(reader.read_bool().unwrap());
        assert_eq!(reader.read_string().unwrap(), "hi");
        let values = reader.read_vec(|r| r.read_string()).expect("read vec");
        assert_eq!(values, vec!["a".to_string(), "b".to_string()]);
        reader.ensure_fully_consumed().expect("fully consumed");
    }

    #[test]
    fn option_map_roundtrip() {
        let flags = [true, false, true];
        let option_map = OptionMap::from_flags(&flags).expect("option map");
        let mut writer = WireWriter::new();
        option_map.write(&mut writer).expect("write option map");
        let bytes = writer.into_bytes();
        let mut reader = WireReader::new(&bytes);
        let decoded = OptionMap::read(&mut reader, flags.len()).expect("read option map");
        assert_eq!(decoded, flags);
    }

    #[test]
    fn option_map_rejects_unknown_bits() {
        let mut writer = WireWriter::new();
        writer.write_u8(0b0000_0100);
        let bytes = writer.into_bytes();
        let mut reader = WireReader::new(&bytes);
        let err = OptionMap::read(&mut reader, 2).unwrap_err();
        assert!(err.to_string().contains("unknown bits"));
    }

    #[test]
    fn read_bool_rejects_invalid_value() {
        let mut writer = WireWriter::new();
        writer.write_u8(2);
        let bytes = writer.into_bytes();
        let mut reader = WireReader::new(&bytes);
        let err = reader.read_bool().unwrap_err();
        assert!(err.to_string().contains("Invalid boolean"));
    }
}
